avocado/plugins/diff.py

```diff
@@ -0,0 +1,387 @@
+                job1_results.extend(variants1)
+                job2_results.extend(variants_header)
+                job2_results.extend(variants2)
+
+        if (('all' in args.diff_filter or 'results' in args.diff_filter) and
+           '-results' not in args.diff_filter):
+
+            results1 = []
+            for test in job1_data['tests']:
+                test_result = '%s: %s\n' % (test['url'], test['status'])
+                results1.append(test_result)
+            results2 = []
+            for test in job2_data['tests']:
+                test_result = '%s: %s\n' % (test['url'], test['status'])
+                results2.append(test_result)
+
+            if self._has_diff(results1, results2):
+                test_results_header = ['\n', 'TEST RESULTS\n']
+                job1_results.extend(test_results_header)
+                job1_results.extend(results1)
+                job2_results.extend(test_results_header)
+                job2_results.extend(results2)
+
+        if (('all' in args.diff_filter or 'config' in args.diff_filter) and
+           '-config' not in args.diff_filter):
+
+            config1 = self._get_config(job1_dir)
+            config2 = self._get_config(job2_dir)
+
+            if self._has_diff(config1, config2):
+                config_header = ['\n', 'AVOCADO SETTINGS\n']
+                job1_results.extend(config_header)
+                job1_results.extend(config1)
+                job2_results.extend(config_header)
+                job2_results.extend(config2)
+
+        if (('all' in args.diff_filter or 'sysinfo' in args.diff_filter) and
+           '-sysinfo' not in args.diff_filter):
+
+            sysinfo_pre1 = self._get_sysinfo(job1_dir, 'pre')
+            sysinfo_pre2 = self._get_sysinfo(job2_dir, 'pre')
+
+            if self._has_diff(sysinfo_pre1, sysinfo_pre2):
+                sysinfo_header_pre = ['\n', 'SYSINFO PRE\n']
+                job1_results.extend(sysinfo_header_pre)
+                job1_results.extend(sysinfo_pre1)
+                job2_results.extend(sysinfo_header_pre)
+                job2_results.extend(sysinfo_pre2)
+
+            sysinfo_post1 = self._get_sysinfo(job1_dir, 'post')
+            sysinfo_post2 = self._get_sysinfo(job2_dir, 'post')
+
+            if self._has_diff(sysinfo_post1, sysinfo_post2):
+                sysinfo_header_post = ['\n', 'SYSINFO POST\n']
+                job1_results.extend(sysinfo_header_post)
+                job1_results.extend(sysinfo_post1)
+                job2_results.extend(sysinfo_header_post)
+                job2_results.extend(sysinfo_post2)
+
+        job_diff = difflib.unified_diff(job1_results,
+                                        job2_results,
+                                        fromfile=job1_id,
+                                        tofile=job2_id)
+        if self.term.enabled:
+            job_diff = self._color_diff(job_diff)
+        sys.stdout.writelines(job_diff)
```

Comment:
Well it's possible to write directly to `stdout`, but it's highly discouraged. One should always use the `logging` and for app logging one should use `logging.getLogger("avocado.app")`. To avoid coloring the output, you need to use `log.debug`.

Also despite my 16GB or memory I like to keep programs efficient. This approach is not really efficient as you first fit everything in memory, then you feed it to another memory and in the end `less` puts it again into memory through pipe. I think it'd be better to use the generators and feed everything directly to `less` by logging each section's diff separately.